cloudinit/sources/DataSourceCloudStack.py

```diff
@@ -93,6 +93,60 @@ def __init__(self, sys_cfg, distro, paths):
         self.metadata_address = "http://%s/" % (self.vr_addr,)
         self.cfg = {}
 
+    def _get_domainname(self):
+        """
+        Try obtaining a "domain-name" DHCP lease parameter:
+        - From systemd-networkd lease
+        - From dhclient lease
+        """
+        LOG.debug("Try obtaining domain name from networkd leases")
+        domainname = dhcp.networkd_get_option_from_leases("DOMAINNAME")
+        if domainname:
+            return domainname
+        LOG.debug(
+            "Could not obtain FQDN from networkd leases. "
+            "Falling back to ISC dhclient"
+        )
+
+        lease_file = dhcp.IscDhclient.get_latest_lease()
```

Comment:
@phsm I'd rather see us remove all dependency on IscDhclient in the cloudstack datasource if consensus is that it was never used, rather than adding new dependencies on it.

Thoughts?

Reply:
Hi Brett,

Thanks for your elaborated review. 

Let me first explain Cloudstack's network topology and where the metadata server (aka virtual router) is located in it. That will help us to figure the most elegant solution.
Cloudstack offers two types of networks: shared and isolated.

**Shared network:**
- VMs are bridged into a vlan together with a subnet router that is not controlled by cloudstack: a physical router or an L3 switch.
- The virtual router is just another VM running in the same network, it gets a random free IP address, and acts as a DHCP server and a metadata server.
- This variant is typically used to give your VMs external IPs (VPS).

**Isolated network:**
- virtual router stands between the "public" network and is a gateway for "guest" network.
- Users are able to assign public IPs to the virtual router, configure port-forwardings/NATs/firewall rules etc.

See the attached diagram.
![Cloudstack Networks difference](https://github.com/canonical/cloud-init/assets/9035611/7cdad26e-6314-4d68-ab24-fa94c2c8a2eb)


Knowing this, it becomes apparent that there's a challenge to discover metadata server in case of shared network: the metadata server has an unpredictable IP. 
The only constant, however, is that metadata server is always a DHCP server, both in shared and isolated network cases.

**IscDhclient dilemma**
In order to discover metadata server, cloud-init needs to know what is the IP of the DHCP server in the network.
That what it does now:
1. Tries to resolve domain "data-server". Works only if virtual-router is used as resolver on the VM.
2. Tries to parse systemd-networkd lease file (even though the lease file explicitly says "do not parse").
3. Tries to discover and parse dhclient lease file in well-known locations (works not only for isc-dhclient, but also for freebsd's dhclient as they share the same format).
4. As the last resort, it assumes the default gateway is also the metadata server.

If we want to strike out parsing networkd/dhclient leases, then I'd consider writing a tiny dhcp client as a python module and including it in cloud-init. 
It will:
1. Send DHCPDISCOVER message
2. Receive DHCPOFFER message
3. It will abandon further communication with DHCP server, thus will not trip the lease refresh time.
4. Store the DHCPOFFER parameters and options into some data fields.


Provided the module will be developed on pure python, it should be very much cross-platform.
I could try implementing it.

What do you think?

Reply:
Thanks, @phsm, for the background on the cloudstack's design. Agreed, dhcp is still required.

> If we want to strike out parsing networkd/dhclient leases

To be clear, my argument was that since cloud-init never actually uses dhclient on cloudstack, the dhclient code shouldn't not needed and we shouldn't add more.

That said, I'm going back on that argument: I suppose the initramfs or NetworkManager might have called dhclient. Those are possible causes of dhclient leases on the filesystem today, so I suppose we _should_ support dhclient as long as it is around.

This does expose a gap, however.

Users that are using modern NetworkManager will always fall back to the default gateway, since it doesn't support the same leases as dhclient and systemd. This can probably be worked around by a user with NetworkManager's [`dhcp=dhclient` config](https://networkmanager.dev/docs/api/latest/NetworkManager.conf.html) for now, as long as dhclient gets shipped with their system.

Reply:
@phsm Given my previous comment, I'm happy to land this if we can get the inline comments addressed and some test coverage.